### ISSUE TITLE
Training loss diff with different te versions

### ISSUE THREAD

**Issue by skydoorkai** [NVIDIA/TransformerEngine#2441]
**Describe the bug**

When training a small moe model on Hopper using blockwise fp8, we found that loss diff is large https://github.com/NVIDIA/TransformerEngine/issues/new/choosewhen using different te versions.
The first 10 training steps produce nearly the same losses, then it will diverge. The loss diff is :

te v2.9 vs v2.7: 0.45% loss diff

main branch commit 3d76218ee75626f7d749025d8d877468547f78c6 vs v2.7: 2.2% loss diff

<img width="2208" height="574" alt="Image" src="https://github.com/user-attachments/assets/b42860d3-af22-4a2e-a066-cd192f56a49a" />

ngc docker image 25.08 is used.

What is the reason for this large loss diff?

**Steps/Code to reproduce bug**

Please list *minimal* steps or code snippet for us to be able to reproduce the bug.

A  helpful guide on on how to craft a minimal bug report  http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports. 

**Expected behavior**

Very small loss diff.

**Environment overview (please complete the following information)**

 - Environment location: [Bare-metal, Docker, Cloud(specify cloud provider - AWS, Azure, GCP, Collab)]
 - Method of Transformer Engine install: [pip install or from source]. Please specify exact commands you used to install.
 - If method of install is [Docker], provide `docker pull` & `docker run` commands used

**Environment details**

If NVIDIA docker image is used you don't need to specify these.
Otherwise, please provide:
- OS version
- PyTorch version
- Python version
- Transformer Engine version
- CUDA version
- CUDNN version

**Device details**
- GPU model

**Additional context**

**Comment by ptrendx** [comment]
@skydoorkai Could you provide more information on the kind of model you tried that produced this issue? Ideally if you could provide an example script that would show this behavior that would be very useful. Adding @negvet.

**Comment by ptrendx** [comment]
Also, @skydoorkai I assume that everything else stayed the same between those 2 experiments (container, libraries versions, etc.)? Could you say which of the loss curves comes from which TE version?

**Comment by skydoorkai** [comment]
For 2.2% diff, it is larger due to https://github.com/NVIDIA/TransformerEngine/pull/2266/, which supports fp8 cuda graph, but Megatron I used did not include the required change for it. Fixing it will get loss similar to v2.9.

Did not find the cause between v2.9 and v2.7 yet.

**Comment by skydoorkai** [comment]
After fixing fp8 cuda graph issue, current te version's loss diff with v2.7 is 0.1%, which is reasonable as some module/kernel implementations have be changed. Close this issue.
